src/day8/re_frame/test.cljc

```diff
@@ -1,66 +1,242 @@
+                                                  (rf/remove-post-event-callback cb-id)
+                                                  (reset! (:now-waiting-for test-context) nil)
+                                                  (done))
+
+                                                (ok-pred event)
+                                                (do
+                                                  (rf/remove-post-event-callback cb-id)
+                                                  (reset! (:now-waiting-for test-context) nil)
+                                                  (binding [*test-context* test-context]
+                                                    (callback event))
+                                                  (when (= (:wait-for-depth test-context)
+                                                           @(:max-wait-for-depth test-context))
+                                                    ;; `callback` has completed with no `wait-for*`
+                                                    ;; calls, so we're not waiting for anything
+                                                    ;; further.  Given that `wait-for*` calls are
+                                                    ;; only valid in tail position, the test must
+                                                    ;; now be finished.
+                                                    (done)))
+
+                                                ;; Test is not interested this event, but we still
+                                                ;; need to wait for the one we *are* interested in.
+                                                :else
+                                                nil)))
+      (reset! (:now-waiting-for test-context) ok-ids))))
+
+
+(defmacro wait-for
+  "Execute `body` once an event identified by the predicate(s) `ids` has been handled.
+
+  `ids` and `failure-ids` are means to identify an event. Normally, each would
+  be a simple keyword or a set of keywords.  If an event with event-id of (or
+  in) `ids` is handled, the test will continue by executing the body. If an
+  event with an event-id of (or in) `failure-ids` is handled, the test will
+  abort and fail.
+
+  IMPORTANT NOTE: due to the way async tests in re-frame work, code you want
+  executed after the event you're waiting for has to happen in the `body` of the
+  `wait-for` (in an implicit callback), not just lexically after the the
+  `wait-for` call. In practice, this means `wait-for` must always be in a tail
+  position.
+
+  Eg:
+      (run-test-async
+        (dispatch [:get-user 2])
+        (wait-for [#{:got-user} #{:no-such-user :system-unavailable} event]
+          (is (= (:username @(subscribe [:user])) \"johnny\")))
+        ;; Don't put code here, it will run *before* the event you're waiting
+        ;; for.
+        )"
+  [[ids failure-ids event-sym :as argv] & body]
+  (let [[failure-ids event-sym] (case (count argv)
```

Comment:
This is nice. At some point it could be cool to look at spec for this, but no need now.